stream settings: Include section for right panel in the URL

evykassirer opened the pull request:
Fixes #21017.

evykassirer committed:
stream settings: Open row with function call instead of simulated click.

evykassirer committed:
stream settings: Include section for right panel in the URL.

Fixes #21017.

evykassirer commented on web/src/hash_parser.ts:
```diff
@@ -15,6 +15,18 @@ export function get_hash_section(hash?: string): string {
     return parts[1] || "";
 }
 
+export function get_current_nth_hash_section(n: number): string {
```
I chose to do 1-indexing instead of 0-indexing, but happy to change it if 0 feels more intuitive to other folks

timabbott replied:
This is probably mostly fine for now

evykassirer commented on web/src/stream_edit.js:
```diff
@@ -39,8 +39,20 @@ import * as user_groups from "./user_groups";
+    ["personal_settings", "personal"],
+    ["subscriber_settings", "subscribers"],
+    ["general_settings", "general"],
+]);
```
It would be cool to have a shared data structure instead of writing out the pairs twice like this, but I couldn't think of something nice to do.

timabbott replied:
I think the better plan would be to just make the URL fragment names be the same as the code-level tab names.

evykassirer replied:
Not sure which direction to take this. Asked on CZO [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/stream.20settings.20url/near/1681645)

evykassirer replied:
Decided to use the shorter strings (e.g. `general`) for both.

evykassirer commented on web/src/stream_edit.js:
```diff
@@ -39,8 +39,20 @@ import * as user_groups from "./user_groups";
+    ["general_settings", "general"],
+]);
+
+function setup_subscriptions_stream_hash(sub, right_section) {
```
`right_section` can sometimes be not passed to this function, is it better to explicitly do `right_section=null` or something in those cases?

evykassirer replied:
(same question for some other functions in this call stack)

timabbott replied:
Better might be to make `hash_util.stream_edit_url` expect a string always, and just pass `"general"` for this second parameter here?

evykassirer replied:
Yeah, I can explicitly pass "general" for all the other function calls

evykassirer commented on web/src/stream_edit.js:
```diff
@@ -430,6 +442,13 @@ export function initialize() {
+        const url_fragment = right_section_name_to_url_fragment.get(tab);
+        const sub = get_sub_for_target(e.target);
+        setup_subscriptions_stream_hash(sub, url_fragment);
+    });
```
I wanted to put this within the existing tab switcher logic, but it's wrapped up in `stream_edit_toggler` and it was hard to put the relevant code in `callback` there, both for typescript and circular import reasons. So I just put it here instead.

timabbott replied:
Can you be specific about what problem you encountered? In general, I would prefer to see code with circular imports and help you figure out how to fix those than to see code that is perturbed from how you think is most readable trying to work around them. It is generally not hard to good write code without circular imports.

evykassirer replied:
I'll try to remember to write out more thoroughly the issues I'm having in the future, especially since most of the time when I do that I fix the problem myself. Which is what I just did :p Pushing with the code now in `stream_edit_toggler`

timabbott commented:
@sahil839 can you review this one? @m-e-l-u-h-a-n might also have some thoughts from working on this part of the code a lot.

timabbott commented on web/src/hashchange.js:
```diff
@@ -258,7 +258,9 @@ function do_hashchange_overlay(old_hash) {
     if (coming_from_overlay && base === old_base) {
         if (base === "streams") {
-            stream_settings_ui.change_state(section);
+            // e.g. #streams/29/social/subscribers
```
Hmm, `/29-social/` is the convention we use for streams in `#narrow` URLs, rather than `/29/social/. It's probably not worth much time to change this now with other priorities, but it might be worth opening an issue for it.

timabbott commented on web/src/hashchange.js:
```diff
@@ -323,7 +325,9 @@ function do_hashchange_overlay(old_hash) {
-        stream_settings_ui.launch(section);
+        // e.g. #streams/29/social/subscribers
+        const right_section = hash_parser.get_current_nth_hash_section(4);
+        stream_settings_ui.launch(section, right_section);
```
`right_side_tab` might be a better name than `right_section` for the selected tab.

evykassirer replied:
It was named alongside the `left_section` variable, would you like to rename that to `left_side_tab` as well? I'm going to leave it as is for now since I think it's better to have `left_section` and `right_section` than `left_section` and `right_side_tab` (makes the left variable seem like it's more than just the tab)

timabbott replied:
Yes, that's probably better. Though I will note that the `left_section/left_side_tab` variable that this PR introduces is somewhat incoherent -- the previous singleton `section` variable could be "new", which is OK, but "new" is a state of the right side, not of the left side (and I think we just preserve which tab on the left side is open?).

I think we want to pass `"subscribed", "new"` after parsing `new` from the URL to correct that, since it looks like `"new"` is always shown with the "subscribed" tab on the left panel.

evykassirer replied:
Oh, I forgot I'd renamed `left_section` 🤦 sorry about that. I can rename them both to `X_side_tab`.

evykassirer replied:
`"new"` isn't shown with "subscribed" in the left panel if the popover is already open and in the "all streams" section

It might actually make sense to keep it as `section` and have the right side tab sometimes null? Because the `section` can also sometimes be a stream identifier.

timabbott replied:
That's probably workable, but we should open a follow-up issue; fundamentally, the `new` state is about the right panel, not the left one.

evykassirer replied:
Opened an issue #27730

timabbott commented on web/src/stream_settings_ui.js:
```diff
@@ -688,7 +689,15 @@ export function switch_to_stream_row(stream_id) {
-    stream_edit.open_edit_panel_for_row($stream_row);
+    stream_edit.open_edit_panel_for_row($stream_row, right_section);
+    const right_section_name = stream_edit.right_section_url_fragment_to_section_name.get(
+        right_section ?? "general",
```
What's the thinking behind the `general` default here?

evykassirer replied:
Changed it so the caller always passes the correct value

evykassirer committed:
stream subscribers: Name the tabs with a single word, removing _settings.

This is preparation for #27637, where it will be nice to be able
to be able to parse a url such as `/#streams/1/announce/general`
and use the `general` string to directly open a stream subscription
tab without converting it to `general_settings`.

evykassirer commented:
Made comments from changes and/or asked followup questions, pushing with the changes noted above.

evykassirer commented:
ready for re-review

timabbott commented on web/templates/stream_settings/stream_settings.hbs:
```diff
@@ -29,7 +29,7 @@
     <div class="inner-box">
 
-        <div class="general_settings stream_section">
+        <div class="general stream_section">
```
Does this class name need to match the name of the tab in in the toggle widget? Because these new class names are awful, unscoped names. A good choice to migrate to would be `stream_section_general`.

timabbott replied:
I at least don't see any obvious thing that requires a class whose name is the same as the `key` for a toggle widget to be applied to the HTML for that element.

evykassirer replied:
It does need to be the same or else explicitly mapped

```
export function setup_toggler(): void {
...
        callback(_name, key) {
            $(".stream_section").hide();
            $(`.${CSS.escape(key)}`).show();
```

That last line is what shows the body of the tab, and the classname needs to match the `key`.

@timabbott  Would you prefer I set up a map here to map from e.g. "stream_section_personal" to just "personal"? Would you prefer to keep it as is? Something else?

timabbott replied:
I made the following change to this commit in order to avoid the awful unscoped class names:

``` diff
diff --git a/web/src/stream_edit_toggler.ts b/web/src/stream_edit_toggler.ts
index eded11f0c6..2866b12173 100644
--- a/web/src/stream_edit_toggler.ts
+++ b/web/src/stream_edit_toggler.ts
@@ -16,7 +16,7 @@ export function setup_toggler(): void {
         ],
         callback(_name, key) {
             $(".stream_section").hide();
-            $(`.${CSS.escape(key)}`).show();
+            $(`[data-stream-section="${CSS.escape(key)}"]`).show();
             select_tab = key;
         },
     });
diff --git a/web/templates/stream_settings/stream_settings.hbs b/web/templates/stream_settings/stream_settings.hbs
index 4acfd435ad..58483628af 100644
--- a/web/templates/stream_settings/stream_settings.hbs
+++ b/web/templates/stream_settings/stream_settings.hbs
@@ -29,7 +29,7 @@
 <div class="subscription_settings" data-stream-id="{{sub.stream_id}}">
     <div class="inner-box">
 
-        <div class="general stream_section">
+        <div class="stream_section" data-stream-section="general">
             {{#with sub}}
             <div class="stream-settings-tip-container">
                 {{> stream_settings_tip}}
@@ -94,7 +94,7 @@
             </div>
         </div>
 
-        <div id="personal-stream-settings" class="personal stream_section">
+        <div id="personal-stream-settings" class="stream_section" data-stream-section="personal">
             <div class="subsection-header">
                 <h3 class="stream_setting_subsection_title inline-block">{{t "Personal settings" }}</h3>
                 <div id="stream_change_property_status{{sub.stream_id}}" class="stream_change_property_status alert-notification"></div>
@@ -141,7 +141,7 @@
             </div>
         </div>
 
-        <div class="subscribers stream_section">
+        <div class="stream_section" data-stream-section="subscribers">
             {{#with sub}}
             <div class="edit_subscribers_for_stream">
                 {{> stream_members}}

```

timabbott commented on web/src/stream_settings_ui.js:
```diff
@@ -792,7 +794,7 @@ export function switch_rows(event) {
+        switch_to_stream_row(stream_id, "general");
     } else if (event === "up_arrow" && !row_data) {
         $("#search_stream_name").trigger("focus");
     }
```
If I visit http://localhost:9991/#streams/12/support/subscribers, and repeatedly click on that stream's row in the left side, it very clearly has the URL change twice each time.

evykassirer replied:
Ah okay so there's a few things going on.

---

One is that the url is updating from two places when clicking a stream on the left side:

<img width="576" alt="image" src="https://github.com/zulip/zulip/assets/5634097/19667bde-6a1a-4d93-adfa-0a69d1654716">

I tried removing the first call, but then the second call never happens because it gets cut off here because the url hasn't changed yet.

```js
export function update_toggler_for_sub(sub) {
    if (!hash_parser.is_editing_stream(sub.stream_id)) {
        return;
    }
```

And we definitely need the second call because that's what updates the url when a user clicks on the tabs themselves to change it.

So I think it's fine to just keep both of them.

---

But we do want them to call the same thing! `open_edit_panel_for_row(this, "general");` was causing the general tab to open, but then `update_toggler_for_sub` uses a pre-saved `stream_edit_toggler.select_tab` instead. We could pick either, but I decided to go with existing logic from before this PR and always open the `stream_edit_toggler.select_tab`. So here's what I ended up changing to fix this in the end:

```diff
diff --git a/web/src/stream_edit.js b/web/src/stream_edit.js
index 1f20e345c5..ad3e62df5b 100644
--- a/web/src/stream_edit.js
+++ b/web/src/stream_edit.js
@@ -548,7 +548,7 @@ export function initialize() {
         stream_settings_components.sub_or_unsub(sub, $stream_row);
 
         if (!sub.subscribed) {
-            open_edit_panel_for_row($stream_row, "general");
+            open_edit_panel_for_row($stream_row, stream_edit_toggler.select_tab);
         }
         stream_ui_updates.update_regular_sub_settings(sub);
 
@@ -616,7 +616,7 @@ export function initialize() {
 
     $("#streams_overlay_container").on("click", ".stream-row", function (e) {
         if ($(e.target).closest(".check, .subscription_settings").length === 0) {
-            open_edit_panel_for_row(this, "general");
+            open_edit_panel_for_row(this, stream_edit_toggler.select_tab);
         }
     });
 ```

evykassirer replied:
Pushing with this change now.

timabbott replied:
That looks reasonable to me!

alya commented:
Works for me in manual testing.

evykassirer commented:
Pushed with a change for one of the comments, left a followup question for the other comment.

timabbott commented:
I made the changes noted in https://github.com/zulip/zulip/pull/27637#discussion_r1399848775 and marked this to merge once CI passes, thanks @evykassirer!
